File: src/lib/models/SGDB.ts
/**
 * Copyright (C) 2024 Travis Lane (Tormak)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>
 */
import { requestTimeoutLength } from "@stores/State";
import type { GridResults, SGDBGame, SGDBImageOptions, SGDBOptions } from "@types";


export class RequestError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "Request Error"
  }
}

/**
 * Wrapper for the SteamGridDB API.
 */
export class SGDB {
  private readonly key: string;
  private readonly baseURL: string;
  private readonly headers: Record<string, any>;

  /**
   * Creates a new SGDB instance.
   * @param options The api key or SGDB options.
   */
  constructor(options: SGDBOptions | string) {
    // Allow passing just the API key as a string
    if (typeof options === "string") {
      options = { key: options };
    }

    this.baseURL = options.baseURL ?? "https://www.steamgriddb.com/api/";
    this.key = options.key ?? "";
    this.headers = {};

    if (options.headers) {
      this.headers = Object.assign({}, options.headers);
    }

    if (this.key) {
      this.headers.Authorization = `Bearer ${this.key}`;
      this.headers.Accept = 'application/json';
    } else {
      process.emitWarning("API Key not provided, some methods won't work.");
    }
  }

  /**
   * Helper function to format query paramters based on the provided options.
   * @param options The query's options.
   * @returns The built query.
   */
  private buildQuery(options: any): { [key: string]: string; } {
    const multiParams = ["styles", "dimensions", "mimes", "types", "platformdata"];
    const singleParams = ["nsfw", "humor", "epilepsy", "oneoftag", "page"];
    const params: any = {};

    multiParams.forEach((queryParam) => {
      if (options[queryParam]?.length) {
        params[queryParam] = options[queryParam].join(",");
      }
    });

    singleParams.forEach((queryParam) => {
      if (typeof options[queryParam] !== "undefined") {
        params[queryParam] = options[queryParam];
      }
    });
    
    return params;
  }

  /**
   * General request function for intereacting with the SteamGridDB api.
   * @param method The http method.
   * @param url The api request url.
   * @param params Optional request parameters.
   * @param formData Optional form data.
   * @param usePublic Whether to use the public api or v2.
   * @returns A promise resolving to the request's result.
   */
  async handleRequest(method: "GET", url: string, params: { [key: string]: string; } = {}, formData = null, usePublic = false): Promise<any> {
    let strParams: string;

    if (Object.entries(params).length > 0) {
      strParams = "";
      Object.entries(params).forEach(([param, val]) => {
        strParams = strParams.concat(`&${param}=${val}`);
      });
      strParams = strParams.substring(1);
    } else {
      strParams = "";
    }

    let options = {
      headers: this.headers,
      method,
      signal: AbortSignal.timeout(requestTimeoutLength)
    };

    if (formData) {
      options = Object.assign({}, options, { formData: formData });
    }

    try {
      let response = await fetch(`${this.baseURL}${usePublic ? "public" : "v2"}${url}${strParams ? `?${strParams}` : ""}`, options);

      const data = await response.json();
      if (data.success) {
        return data;
      } else {
        throw new RequestError(data.errors?.join(", ") ?? "Unknown SteamGridDB error.");
      }
    } catch (error: any) {
      throw new RequestError(error.message ?? "SteamGridDB error.");
    }
  }

  /**
   * Gets a list of possible matches for a query.
   * @param query The search query.
   * @returns A promise resolving to a list of possible matches.
   */
  async searchGame(query: string): Promise<SGDBGame[]> {
    return (await this.handleRequest("GET", `/search/autocomplete/${encodeURIComponent(query)}`)).data;
  }

  /**
   * Gets grids for a game given its platform and id.
   * @param options The SGDB request options.
   * @returns A promise resolving to the game's grids.
   */
  async getGrids(options: SGDBImageOptions): Promise<GridResults> {
    const results = await this.handleRequest("GET", `/grids/${options.type}/${options.id}`, this.buildQuery(options));
    const grids = results.data.map((img: any) => {
      img.isAnimated = img.thumb.includes('.webm');
      return img;
    });

    return {
      images: grids,
      page: results.page,
      total: results.total,
    };
  }

  /**
   * Gets a list of grids based on the provided game id and filters.
   * @param id The game's id.
   * @param styles List of styles to include.
   * @param dimensions List of dimensions to include.
   * @param mimes List of mimes to include.
   * @param types List of types to include.
   * @param nsfw Whether the result should include nsfw images.
   * @param humor Whether the result should include humor images.
   * @param epilepsy Whether the result should include epilepsy images.
   * @param page The page of results to get.
   * @returns A promise resolving to a list of grids for the desired game matching the provided filters.
   */
  async getGridsById(
    id: number,
    styles?: string[],
    dimensions?: string[],
    mimes?: string[],
    types?: string[],
    nsfw?: string,
    humor?: string,
    epilepsy?: string,
    page?: number
  ): Promise<GridResults> {
    return this.getGrids({
      type: "game",
      id: id,
      styles: styles,
      dimensions: dimensions,
      mimes: mimes,
      types: types,
      nsfw: nsfw,
      humor: humor,
      epilepsy: epilepsy,
      page: page
    });
  }
}

File: src-rust/src/websocket/sgdb.rs
use std::{collections::HashMap, time::Duration};

use reqwest::{header::{self, HeaderMap, HeaderValue}, Client};
use serde::{de::DeserializeOwned, Deserialize};

use super::types::sgdb::{GridResults, SGDBGame, SGDBImage};

#[derive(Clone, Deserialize)]
struct ResponseGeneric<T> {
  pub data: T,
}

#[derive(Clone, Deserialize)]
struct Images {
  images: Vec<SGDBImage>,
}

#[derive(Clone, Deserialize)]
struct Results {
  pub data: Images,
  pub page: u64,
  pub total: u64,
}

#[derive(Clone)]
pub struct SGDBClient {
  key: String,
  base_url: String,
  client: Client,
}

impl SGDBClient {
  /// Creates a new SGDB Client
  pub fn new(key: String, timeout: u64) -> SGDBClient {
    let mut headers = HeaderMap::new();

    headers.insert(header::AUTHORIZATION, HeaderValue::try_from(format!("Bearer {key}")).unwrap());
    headers.insert(header::ACCEPT, HeaderValue::from_static("application/json"));

    let http_client_res = Client::builder()
      .timeout(Duration::from_secs(timeout))
      .default_headers(headers)
      .build();
    let http_client: Client = http_client_res.expect("Failed to make the reqwest client.");

    return SGDBClient {
      key,
      base_url: "https://www.steamgriddb.com/api/".to_string(),
      client: http_client,
    };
  }

  /// Makes a request.
  async fn handle_request<T: DeserializeOwned>(self, url: String, params: Option<HashMap<String, String>>) -> Result<T, String> {
    let mut stringified_params: String = "".to_string();

    if params.is_some() {
      params.unwrap().into_iter().for_each(| (param, val) | {
        let entry = format!("&{param}={val}");
        stringified_params.push_str(&entry);
      });
      let without_mark = (&stringified_params[1..]).to_string();
      
      stringified_params = "?".to_string();
      stringified_params.push_str(&without_mark);
    }
    
    let response_res = self.client.get(format!("{}v2{}{}", self.base_url, url, stringified_params)).send().await;

    if response_res.is_ok() {
      let response = response_res.ok().expect("Failed to get response from ok result.");
      let data: T = response.json().await.expect("Data should have been of type");
  
      return Ok(data);
    } else {
      let err = response_res.err().expect("Request failed, error should have existed.");
      return Err(err.to_string());
    }
  }


  /// Searches for SGDB games by their id.
  pub async fn search_game(self, query: String) -> Result<Vec<SGDBGame>, String> {
    let encoded_query = urlencoding::encode(&query).into_owned();

    let res = self.handle_request::<ResponseGeneric<Vec<SGDBGame>>>(format!("/search/autocomplete/{encoded_query}"), None).await;
    if res.is_err() {
      return Err(res.err().unwrap());
    }

    return Ok(res.unwrap().data);
  }

  /// Gets the grids for a game based on its SGDB id.
  async fn get_grids_by_id(self, id: u64, page: u64) -> Result<GridResults, String> {
    let mut params = HashMap::new();

    params.insert("types".to_string(), "static,animated".to_string());
    params.insert("nsfw".to_string(), "any".to_string());
    params.insert("humor".to_string(), "any".to_string());
    params.insert("epilepsy".to_string(), "any".to_string());
    params.insert("page".to_string(), page.to_string());

    let res = self.handle_request::<Results>(format!("/grids/game/{id}"), Some(params)).await;
    if res.is_err() {
      return Err(res.err().unwrap());
    }
    let results = res.unwrap();

    let grids = results.data.images.into_iter().map(| mut img | {
      img.isAnimated = img.thumb.contains(".webm");
      return img;
    }).collect();

    return Ok(GridResults {
      images: grids,
      page: results.page,
      total: results.total,
    });
  }
}